pennylane/workflow/qnode.py

```diff
@@ -96,13 +99,92 @@ def _make_execution_config(
 
     return qml.devices.ExecutionConfig(
         interface=getattr(circuit, "interface", None),
-        gradient_method=_gradient_method,
+        gradient_keyword_arguments=gradient_kwargs,
+        gradient_method=diff_method,
         grad_on_execution=grad_on_execution,
         use_device_jacobian_product=execute_kwargs.get("device_vjp", False),
         mcm_config=mcm_config or qml.devices.MCMConfig(),
     )
 
 
+def _update_mcm_config(mcm_config: "qml.devices.MCMConfig", interface: str, finite_shots: bool):
+    """Helper function to update the mid-circuit measurements configuration based on
+    execution parameters"""
+    if interface == "jax-jit" and mcm_config.mcm_method == "deferred":
+        # This is a current limitation of defer_measurements. "hw-like" behaviour is
+        # not yet accessible.
+        if mcm_config.postselect_mode == "hw-like":
+            raise ValueError(
+                "Using postselect_mode='hw-like' is not supported with jax-jit when using "
+                "mcm_method='deferred'."
+            )
+        mcm_config.postselect_mode = "fill-shots"
+
+    if (
+        finite_shots
+        and "jax" in interface
+        and mcm_config.mcm_method in (None, "one-shot")
+        and mcm_config.postselect_mode in (None, "hw-like")
+    ):
+        mcm_config.postselect_mode = "pad-invalid-samples"
+
+
+def _resolve_execution_config(
+    tapes: QuantumScriptBatch,
+    device: "qml.devices.Device",
+    execution_config: "qml.devices.ExecutionConfig",
+    transform_program: TransformProgram,
+) -> "qml.devices.ExecutionConfig":
+    """Resolves the execution configuration for non-device specific properties.
+
+    Args:
+        tapes (QuantumScriptBatch): a batch of tapes
+        device (qml.devices.Device): a Pennylane device
+        execution_config (qml.devices.ExecutionConfig): an execution config to be executed on the device
+        transform_program (TransformProgram): a program of transformations to be applied to the tapes
+
+    Returns:
+        qml.devices.ExecutionConfig: resolved execution configuration
+    """
+    if (
+        device.name == "lightning.qubit"
+        and qml.metric_tensor in transform_program
+        and execution_config.gradient_method == "best"
+    ):
+        gradient_fn = qml.gradients.param_shift
+    else:
+        gradient_fn = QNode.get_gradient_fn(
+            device, execution_config.interface, execution_config.gradient_method, tape=tapes[0]
+        )[0]
+
+    if gradient_fn is qml.gradients.param_shift_cv:
+        execution_config.gradient_keyword_arguments["dev"] = device
+
+    execution_config.gradient_method = gradient_fn

```

Comment:
We should probably also be using `dataclasses.replace` or first copying the execution config so that we are not mutating the input to this function.